Fix compiler crash (#891)

srdjanstipic opened the pull request.

srdjanstipic commented:
I am aware that this PR does not include the test.
Where (in which file/test) should I put the crashing program?

SeanTAllen commented:
Re: tests.

You can add a test as part of the compiler test suite.

How large is your minimal test case?

srdjanstipic commented:
@SeanTAllen Is './test/libponyc/badpony.cc' a good place for the test?
The minimal test is:

```ponyc
primitive T

actor @A
  fun f(a: T): (T | None) =>
    a
```

SeanTAllen commented:
@srdjanstipic i'm not sure if badpony.cc is the correct location, but its better than nothing and its a good place to start. @jemc or @sylvanc might have a better idea where it should go.

srdjanstipic commented:
I have made the push with updated commit message and updated CHANGELOG.md.
I am aware that this is not the final PR because the test that I added in badpony.cc is not correct.
Test is not correct because it does not fail without the fix.
I do not know how to generate a test that calls genheader().
Comments are welcomed.

sylvanc commented:
I think we should merge this even though there isn't a "proper test" yet. It fixes a problem, and, to be honest, I'm not sure how to generate a test that calls `genheader` either!
